src/i_sdlmusic.c

```diff
@@ -645,18 +646,42 @@ static boolean ReadSubstituteConfig(char *filename)
         return false;
     }
 
-//    old_subst_music_len = subst_music_len;
+    length = M_FileLength(fs);
 
-    while (!feof(fs))
+    buffer = malloc(length + 1);
+    if (buffer == NULL)
     {
-        char *retval;
-        M_StringCopy(line, "", sizeof(line));
-        retval = fgets(line, sizeof(line), fs);
+        fclose(fs);
+        I_Error("Out of memory while trying to read music substitution config \"%s\" (%ld bytes)", filename, length);
+    }
+
+    retval = fread(buffer, 1, length, fs);
+    if (retval != length) {
+        free(buffer);
+        fclose(fs);
+        I_Error("Error while reading music substitution config \"%s\" (expected %ld, got %ld bytes)", filename, length, retval);
+    }
+    buffer[length] = '\0';
+    line = buffer;
 
-        if (retval == NULL)
+    while (*line != '\0')
+    {
+        char *error;
+        char *next;
+
+        // find end of line
+        char *eol = strchr(line, '\n');
+        if (eol != NULL)
         {
-            fprintf(stderr, "%s:%i: Unexpected end of file\n", filename, linenum);
-            break;
+            // change the newline into NUL
+            *eol = '\0';
+            next = eol + 1;
+        }
+        else
+        {
+            // end of buffer
+            eol = buffer + length;
```

Comment:
Setting `eol` isn't necessary. Just `next = NULL;` and check for `while (line != NULL)`.